The purpose of this notebook is to understand the basics of neural networks.

We will start with a single neuron and then move on to a simple neural network.





In [178]:
import torch
import pandas as pd

torch.manual_seed(0)


class Model(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = torch.nn.Linear(1, 1)

    def forward(self, x):
        # Pass through the linear layer (single neuron)
        return self.linear(x)


# Create an instance of the model
model = Model()
print(f"{model=}")

model=Model(
  (linear): Linear(in_features=1, out_features=1, bias=True)
)


In [179]:
# Let's examine the model parameters
print("Model parameters:")
for name, param in model.named_parameters():
    print(f"{name}: {param.data}")
print(f"{model(torch.tensor([1.0]))=}")


Model parameters:
linear.weight: tensor([[-0.0075]])
linear.bias: tensor([0.5364])
model(torch.tensor([1.0]))=tensor([0.5290], grad_fn=<ViewBackward0>)


The output of the model for the input of 1.0 is a tensor with a single value that is the result of a random initialization of the weights and biases.
The output is 0.5290, which is input * weight + bias: 1.0 * -0.0075 + 0.5364 = 0.5289. This is slightly different than 0.5290 b/c the printed values are rounded from the actual float values used.

By default, autograd is tracking the operations that are performed on the model.
`grad_fn=<ViewBackward0>` tells us that the last operation to create this output was a view operation. Without the view operation the output would have been [[0.5290]] rather than [0.5290]. The `Backward` means that during backgpropagation the Backward function will be called. The `0` suffix means that this will be the first backward call.







In [180]:
# Let's create a simple training example
def train_step(x: torch.Tensor, y: torch.Tensor) -> None:
    criterion = torch.nn.MSELoss()

    # Forward pass
    output = model(x)
    loss = criterion(output, y)
    optimizer = torch.optim.SGD(model.parameters(), lr=0.1)

    # Backward pass
    loss.backward()  # This computes gradients for all model parameters
    optimizer.step()  # This uses the gradients to update the weights and biases
    weight_grad = model.linear.weight.grad.item()
    bias_grad = model.linear.bias.grad.item()
    # Reset gradients for the next step
    optimizer.zero_grad()
    
    # Return values for tabulation
    return {
        'prediction': output.item(),
        'loss': loss.item(),
        'weight': model.linear.weight.data.item(),
        'bias': model.linear.bias.data.item(),
        'weight_grad': weight_grad,
        'bias_grad': bias_grad
    }
first_result = train_step(x=torch.tensor([1.0]), y=torch.tensor([1.0]))
first_result

{'prediction': 0.5289567708969116,
 'loss': 0.2218817174434662,
 'weight': 0.08672183007001877,
 'bias': 0.6306522488594055,
 'weight_grad': -0.9420864582061768,
 'bias_grad': -0.9420864582061768}

Here, the loss is ~0.22 which is calculated from the MSE: (1-0.5289)^2 = 0.2219.  
The new gradients are the result of the optimizer updating the weights and biases.  
For MSELoss, the gradient is ∂Loss/∂output = 2 * (output - target) / n, which is 2 * (0.5289 - 1) / 1 = -0.9422 for the weight and bias (not shown). This is then used to update the weights and biases from above:  
new weight = old_weight - learning_rate * gradient_weight = -0.0075 - 0.01 * -0.9422 = 0.0019  
new bias = old_bias - learning_rate * gradient_bias = 0.5364 - 0.01 * -0.9422 = 0.5458


In [181]:

# Create a DataFrame from the first training result
results_df = pd.DataFrame([first_result])
# Add an iteration column and set it to 0
results_df.insert(0, 'iteration', 0)

for i in range(1, 50):
    result = train_step(x=torch.tensor([1.0]), y=torch.tensor([1.0]))
    # Create a new row with the current iteration and result
    new_row = pd.DataFrame([{**{'iteration': i}, **result}])
    # Append the new row to results_df
    results_df = pd.concat([results_df, new_row], ignore_index=True)
results_df

iteration  prediction          loss    weight      bias   weight_grad  \
0           0    0.528957  2.218817e-01  0.086722  0.630652 -9.420865e-01   
1           1    0.717374  7.987741e-02  0.143247  0.687177 -5.652518e-01   
2           2    0.830424  2.875588e-02  0.177162  0.721093 -3.391511e-01   
3           3    0.898255  1.035212e-02  0.197511  0.741442 -2.034907e-01   
4           4    0.938953  3.726760e-03  0.209721  0.753651 -1.220944e-01   
5           5    0.963372  1.341637e-03  0.217046  0.760977 -7.325673e-02   
6           6    0.978023  4.829889e-04  0.221442  0.765372 -4.395401e-02   
7           7    0.986814  1.738763e-04  0.224079  0.768009 -2.637243e-02   
8           8    0.992088  6.259471e-05  0.225661  0.769592 -1.582336e-02   
9           9    0.995253  2.253432e-05  0.226611  0.770541 -9.494066e-03   
10         10    0.997152  8.112289e-06  0.227180  0.771111 -5.696416e-03   
11         11    0.998291  2.920424e-06  0.227522  0.771452 -3.417850e-03   
12         12    0.998975  1.051401e-06  0.227727  0.771658 -2.050757e-03   
13         13    0.999385  3.785192e-07  0.227850  0.771781 -1.230478e-03   
14         14    0.999631  1.362581e-07  0.227924  0.771854 -7.382631e-04   
15         15    0.999779  4.905820e-08  0.227968  0.771899 -4.429817e-04   
16         16    0.999867  1.765145e-08  0.227995  0.771925 -2.657175e-04   
17         17    0.999920  6.360224e-09  0.228011  0.771941 -1.595020e-04   
18         18    0.999952  2.285120e-09  0.228020  0.771951 -9.560585e-05   
19         19    0.999971  8.253807e-10  0.228026  0.771957 -5.745888e-05   
20         20    0.999983  2.967262e-10  0.228030  0.771960 -3.445148e-05   
21         21    0.999990  1.063292e-10  0.228032  0.771962 -2.062321e-05   
22         22    0.999994  3.842615e-11  0.228033  0.771963 -1.239777e-05   
23         23    0.999996  1.365663e-11  0.228034  0.771964 -7.390976e-06   
24         24    0.999998  5.130119e-12  0.228034  0.771965 -4.529953e-06   
25         25    0.999999  1.719513e-12  0.228034  0.771965 -2.622604e-06   
26         26    0.999999  6.963319e-13  0.228035  0.771965 -1.668930e-06   
27         27    1.000000  2.273737e-13  0.228035  0.771965 -9.536743e-07   
28         28    1.000000  5.684342e-14  0.228035  0.771965 -4.768372e-07   
29         29    1.000000  1.421085e-14  0.228035  0.771965 -2.384186e-07   
30         30    1.000000  1.421085e-14  0.228035  0.771965 -2.384186e-07   
31         31    1.000000  1.421085e-14  0.228035  0.771965 -2.384186e-07   
32         32    1.000000  3.552714e-15  0.228035  0.771965 -1.192093e-07   
33         33    1.000000  3.552714e-15  0.228035  0.771965 -1.192093e-07   
34         34    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
35         35    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
36         36    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
37         37    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
38         38    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
39         39    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
40         40    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
41         41    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
42         42    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
43         43    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
44         44    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
45         45    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
46         46    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
47         47    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
48         48    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   
49         49    1.000000  0.000000e+00  0.228035  0.771965  0.000000e+00   

       bias_grad  
0  -9.420865e-01  
1  -5.652518e-01  
2  -3.391511e-01  
